### DISCUSSION TITLE
Decide on translation file structure

### DISCUSSION THREAD

**Discussion by chimp1984** [bisq-network/bisq2#76] [General]
For Bisq 2 I considered to use a more fine grained/module based resource bundle structure for translation strings.

So there would be a 'common' resource bundle for general strings, a 'offer' resource bundle for offer domain related strings, etc...
This would help to break up one huge global resource bundle as we use in Bisq 1.

Access would be `Res.common.get("cancel")` or `Res.offer.get("offerbook.table.header.market")`

One problem we had already in Bisq 1 is that some strings used initially in one domain/view (e.g. createOffer) will be later reused in a similar view (e.g. takeOffer) and it becomes weird to read "createOffer" keys inside the take offer view. To rename the keys causes extra effort specially for Transifex.

With the more modular approach that problem will be the same.

I am not sure if the more fine grained separation makes things really better.

Another aspect is that some portion of strings can be reused and by that reduce costs for translations, so we should not lightly drop them just because of some fancy refactoring.
On the other hand we do not want to  carry over polluted content (e.g. unclear/outdated keys, duplicates,...).

So I am not sure whats the best approach here, but as we are getting more and more UI content I think it is time for finding a decision.

One option could be to copy over the Bisq 1 resources into one bundle and when a new string is needed we search if that string exist in that bundle and if so we use it. Otherwise we add the string to a new bundle. We could keep the Bisq1 bundle immutable and up to date with Bisq1. Any new changes will get added in the new bundle. So the translation efforts will be limited to new strings.

At release time we can eliminate the not used strings (IDEA is helpful here).

We could also use one global bundle for the new strings instead the module based approach and use the namespace pattern to separate content (which is anyway needed also with the module based approach, just could reduce key depth by one level).

We should also keep in mind the transifex process. I am not much familar with it, so would be great if those who are more familiar (like @ripcurlx) could post their opinions here?

This trade-off problem between modular separation vs. one global collection will also be present for the css file. The huge css in Bisq is also hard to work with and find things. Maybe @ripcurlx can add his opinion for that issue as well.

**Comment by chimp1984** [comment]
I think best option is to use the Bisq 1 resources in an immutable way and additionally add a new resource bundle for Bisq 2. 
If a string gets requested it looks up first the Bisq2 bundle and if not found looks up the Bisq 1 bundle.

If a string gets added the dev should lookup if it exists in Bisq1 bundle and use that if its existing. If not add a new entry in Bisq 2.
If keys are not well matching we can refactor the keys once we get completed and when we remove Bisq 1 bundle and add all to Transifex.

For common strings ("shared" in Bisq1) I suggest to not use any namespace.

**Reply by ripcurlx** [reply]
In general using multiple resource files on Transifex shouldn't make a difference for the translators as they are not visible to the translator, just the keys. To update multiple resource translation files for releases is also not a problem, as it is done via their API by a small script. 

I think in general having a single translation file vs. small ones for each module is not such a big deal for translations. As we don't load modules during runtime we won't bundle them into the modules anyways.

With the huge css file I see it somehow different. It is very hard to maintain as it is right now. Maybe we could look into some CSS pre-processing solution again or at least structure the CSS in a better way like [TCSS (Inverted Triangle CSS)](https://www.freecodecamp.org/news/managing-large-s-css-projects-using-the-inverted-triangle-architecture-3c03e4b1e6df/) or [BEM (Block Element Modifier)](http://getbem.com/introduction/)

**Reply by chimp1984** [reply]
> CSS pre-processing solution

I am a bit skeptical with that as long there is not a very good reason to use it because it comes with new dependency and tooling requirement adding potential supply chain vulnerabilities. Specially dependencies from the UI/design domain are usually built with less security mindset.

> TCSS 

Just looked quickly into it, but that sounds like a reasonable approach. I would like also that we are much more component focussed on Bisq 2.
